src/v/cloud_storage/inventory/report_parser.h

```diff
@@ -0,0 +1,94 @@
+/*
+ * Copyright 2024 Redpanda Data, Inc.
+ *
+ * Use of this software is governed by the Business Source License
+ * included in the file licenses/BSL.md
+ *
+ * As of the Change Date specified in that file, in accordance with
+ * the Business Source License, use of this software will be governed
+ * by the Apache License, Version 2.0
+ */
+
+#pragma once
+
+#include "base/seastarx.h"
+#include "bytes/details/io_allocation_size.h"
+
+#include <seastar/core/iostream.hh>
+
+#include <zlib.h>
+
+namespace cloud_storage::inventory {
+
+using gzip_compressed_t = ss::bool_class<struct gzip_compressed_tag>;
+
+class report_decompression_error : public std::runtime_error {
+public:
+    using runtime_error::runtime_error;
+};
+
+/// Parses newline delimited inventory report, supplied as an input stream. The
+/// report may optionally be GZip compressed. The rows returned by the parser
+/// have newlines stripped off, but no other modifications are performed. In
+/// particular, whitespaces and carriage returns are still present in the
+/// returned row. For CSV files, parsing of the row is not performed and must be
+/// done by the caller.
+class report_parser {
+public:
+    report_parser(const report_parser&) = delete;
+    report_parser& operator=(const report_parser&) = delete;
+
+    report_parser(report_parser&&) = delete;
+    report_parser& operator=(report_parser&&) = delete;
+
+    // The chunk size supplied here controls the following behavior:
+    // 1. The size of data read from the input stream per read_up_to invocation
+    // 2. The max buffer size allocated during chunk decompression if the report
+    // is compressed
+    explicit report_parser(
+      ss::input_stream<char> stream,
+      gzip_compressed_t is_compressed = gzip_compressed_t::no,
+      size_t chunk_size = details::io_allocation_size::max_chunk_size);
+
+    // Returns the next line from the input stream, or nullopt if the stream is
+    // fully consumed.
+    ss::future<std::optional<ss::sstring>> next_line();
+    bool eof() const { return _eof; }
+
+    // High level API for consuming a report. The consumer is supplied with each
+    // row in the stream until EOF. The consumer must return a future;
+    template<typename Consumer>
+    ss::future<> consume(Consumer c) {
+        for (auto row = co_await next_line(); row.has_value();
+             row = co_await next_line()) {
+            co_await c(row.value());
```

Comment:
This will add 2 scheduling points per line, I'm not saying that this will not work or will necessary be a bottleneck. But some performance test could be nice to have, or maybe a metric for the decoding time.

Reply:
To address this I am trying to add a set of additional methods within this PR which can pass around a fragmented vector instead of single rows. It should be more efficient as for 128 KiB chunks we will almost certainly get multiple hundreds of rows per read.

Reply:
I changed the method to work with batches (frag. vectors) of rows instead of a single row at a time. So now we get batches of rows out of the parser (as many as we can parse from the chunk we read off the HTTP stream) and pass the batch to the consumer.

It is worth noting that the consumer still processes the rows one at a time with a `co_await` because each row could potentially belong to a different ntp and they might tip over the flush limit.

It might be possible to avoid that by deciding to flush only after a batch of rows has been processed.

Reply:
changed to do maybe do a flush only after a batch of rows is processed instead of possibly after each row